node/src/transaction_pool/transaction_pool_reducer.rs

```diff
@@ -297,17 +302,48 @@ impl TransactionPoolState {
                 };
 
                 let dispatcher = state.into_dispatcher();
-                if let Some(rpc_action) = rpc_action {
-                    dispatcher.push(rpc_action);
+
+                // TODO: use callbacks
+                match (was_accepted, from_source) {
+                    (true, TransactionPoolMessageSource::Rpc { id }) => {
+                        dispatcher.push(RpcAction::TransactionInjectSuccess {
+                            rpc_id: id,
+                            response: accepted.clone(),
+                        });
+                    }
+                    (true, TransactionPoolMessageSource::Pubsub { id }) => {
+                        dispatcher.push(P2pNetworkPubsubAction::BroadcastValidatedMessage {
+                            message_id: BroadcastMessageId::MessageId { message_id: id },
+                        });
+                    }
+                    (false, TransactionPoolMessageSource::Rpc { id }) => {
+                        dispatcher.push(RpcAction::TransactionInjectRejected {
+                            rpc_id: id,
+                            response: rejected.clone(),
+                        });
+                    }
+                    (false, TransactionPoolMessageSource::Pubsub { id }) => {
+                        dispatcher.push(P2pNetworkPubsubAction::RejectMessage {
+                            message_id: Some(BroadcastMessageId::MessageId { message_id: id }),
+                            peer_id: None,
+                            reason: "Rejected transaction diff".to_owned(),
+                        });
+                    }
+                    (_, TransactionPoolMessageSource::None) => {}
                 }
-                // TODO: libp2p logic already broadcasts everything right now and doesn't
-                // wait for validation, thad needs to be fixed. See #952
+
                 if was_accepted {
-                    dispatcher.push(TransactionPoolAction::Rebroadcast {
-                        accepted,
-                        rejected,
-                        is_local: is_sender_local,
-                    });
+                    match from_source {
```

Comment:
what about an `is_libp2p` method similar to `is_sender_local` for this check to avoid the need to match here? (contents are not used anyway).

Btw I will check later today, but overall the current version looks good